### ISSUE TITLE
Copying Find Artifact from Execution stages, from within the same pipeline results in stages with duplicated ids

### ISSUE THREAD

**Issue by thecodejunkie** [spinnaker/spinnaker#4354]
### Issue Summary:
Copying Find Artifact from Execution stages, from within the same pipeline results in stages with duplicated ids, but with different `matchArtifact` properties

### Cloud Provider(s):
GCP

### Environment:
Spinnaker v1.13.5 hosted on GKE

### Feature Area (if this issue is UI/UX related, please tag `@spinnaker/ui-ux-team`):
Pieplines

### Description:
Copying Find Artifact from Execution stages, from within the same pipeline results in stages with duplicated ids, but with different `matchArtifact` properties. This would mean that you would never get two different artifacts defined in your pipeline. Nothing in the UI suggests that it would overwrite the first artifact.

To me it makes very little sense for the ID to be retrained in the copy, especially with making it crystal clear in the UI. My expectation was that it would copy the other properties, but generate a unique id for it.

### Steps to Reproduce:

1. Create a new pipeline
1. Create a `Find Stage from Execution` stage and fill it out
2. Create a new stage by using the `Copy and existing stage` feature and choose the stage from step 2
3. Inspect JSON from both stages and they will have the same ID

### Additional Details:

Here is the JSON of two steps that I created. The `Find Configmap` was created first and `Find Deployment` is a copy of `Find Configmap`, but with an altered value for the `File path` property

```
Find Configmap

{
  "application": "hello-service",
  "executionOptions": {
    "successful": true
  },
  "expectedArtifact": {
    "id": "905fb773-5c03-44cf-8a5f-8fc55d843c22",
    "matchArtifact": {
      "name": "configmap.yaml",
      "type": "github/file"
    }
  },
  "name": "Find Configmap",
  "pipeline": "41e5a15b-cbc8-4edd-9656-ac0a8d51da6b",
  "type": "findArtifactFromExecution"
}
```

```
Find Deployment

{
  "application": "hello-service",
  "executionOptions": {
    "successful": true
  },
  "expectedArtifact": {
    "id": "905fb773-5c03-44cf-8a5f-8fc55d843c22",
    "matchArtifact": {
      "name": "deployment.yaml",
      "type": "github/file"
    }
  },
  "name": "Find Deployment",
  "pipeline": "41e5a15b-cbc8-4edd-9656-ac0a8d51da6b",
  "type": "findArtifactFromExecution"
}
```

Notice that both have the same `id` value.

**Comment by maggieneterval** [comment]
Hey @thecodejunkie, apologies for the delayed followup! I'm looking into this now, and understand how it might be confusing to end up with two Find Artifact from Execution stages with the same artifact ID, when your intention is to simply use the "Copy an existing stage" functionality as a starting point for a stage meant to find a different artifact. Since you are trying to create a stage to find a completely different artifact, is there a reason you're using the Copy functionality rather than just creating a new stage? I'm guessing to avoid having to re-select the application, pipeline, and kind?

If we were to change the stage copying logic such that certain fields, such as artifact ID, for certain stage types, such as Find Artifact from Execution, were re-initialized, I would think that we would want to be consistent and also re-initialize these fields when copying any stage that defines an artifact ID, which will be rather confusing post artifacts rewrite. So I'm a bit hesitant to change the behavior here, but open to further community input on this issue. Copying @jkschneider who led the charge on the artifacts rewrite to see if he has any additional thoughts!

**Comment by thecodejunkie** [comment]
@maggieneterval Hi!

> is there a reason you're using the Copy functionality rather than just creating a new stage? I'm guessing to avoid having to re-select the application, pipeline, and kind?

Yes, it was out of pure convenience. One that I imagine more people in our organization doing + I don't think it's an unlikely action to be performed by any random Spinnaker user. Currently there is no indication that it will actually have the behaviour that it does. I think the current behavior is more that of a "clone" where I'd expect it to be a carbon copy and I interpreted copy more as a "based on" operation.

Could you please elaborate when the existing behavior would be the desired one? How often do you need multiple artifact stages that all compete for the same artifact id? I guess if you used expressions to branch your pipeline and populate artifact X from source A or from source B depending on the evaluated expression? In my case I needed to find several artifacts (configmap, deployment, service, ingress and docker) from a another pipeline so it thought I'd make my life easier by simply copying one and just altering the name of the artifact it should look for

Thanks!

**Comment by maggieneterval** [comment]
@thecodejunkie thank you so much for elaborating on your user case here, makes sense to me! It sounds like your team is interpreting the "copy stage" functionality as more of "copy stage as a starting point, excluding artifacts" which is understandable. There's not really any precedent for excluding certain stage properties when copying a stage; currently the source stage properties are just copied wholesale. Since the motivation behind this change is convenience, just tossing out a totally different possible solution: what if you could find *multiple* artifacts from the same pipeline execution in a *single* Find Artifact From Execution stage? Not sure how possible that would be to implement but curious for your thoughts since it sounds like you and possibly others often need to find multiple produced artifacts from a given pipeline execution!

**Comment by thecodejunkie** [comment]
@maggieneterval 

1. what if you could find _multiple_ artifacts from the same pipeline execution in a single Find Artifact From Execution stage

Honestly I love this year. Would de-clutter the pipeline visualization quite a lot. Just to give some context on what we currently do. We have a `Deploy To Staging` pipeline that fetches manifests from GitHub and a Docker artifact from a GCR trigger. We then have a `Promote To Production` pipeline that finds a bunch of manifests and the docker image, from the last successful `Deploy to Staging` execution. So right now we need as many `Find Artifact(s) from Execution` stages as the number of resources we want to promote up to production. If we could have a single one, and simply give it a name such as `Find artifacts from staging` then that would be awesome. If this was possible, then it would still be possible to use 1 stage for 1 artifact if you still wanted to or use 1 stage for multiple artifacts

2. if the current behavior for `copy stage` remains then I would suggest that it's made very obvious that the copy would keep the artifactid, just to avoid the confusion. Just by adding a simple text, in the `Copy Stage` dialog window would be huge help and potential time (and headache) saver. Something like this

![image](https://user-images.githubusercontent.com/50543/59284155-0f25c180-8c6c-11e9-8cd1-6b74f02843cb.png)

Right now you never really see the artifact IDs unless you peek under the hood and look at the JSON of the stage definition

**Comment by maggieneterval** [comment]
Thanks for the additional context, @thecodejunkie! Adding my teammates @demobox and @ezimanyi for any additional thoughts on accepting multiple expected artifacts in the Find Artifact From Execution Stage (1), and our technical writer @dorbin for thoughts on the suggested Copy Stage text in (2).

**Comment by ezimanyi** [comment]
It seems completely reasonable to me to expand the "Find Artifact From Execution" stage to allow selecting multiple artifacts from the same pipeline.

**Comment by demobox** [comment]
> It seems completely reasonable to me to expand the "Find Artifact From Execution" stage to allow selecting multiple artifacts from the same pipeline.

+1 to that. I also wonder whether it would make sense to either:

* warn, when duplicating this stage type, that the copied artifact will have the same ID, as suggested above (special case handling, so not ideal)
* remove this stage type from the stage types that can be copied (less convenience for the user, especially if multiple artifacts cannot be selected in a single stage; also special case handling)
* modify the "copy stage" behaviour to (perhaps optionally) not copy any derived/generated properties. This seems to be the root cause of the confusion here - a property is being duplicated that the user isn't aware of just looking at the UI, and so is probably not expecting to be copied

?

**Comment by thecodejunkie** [comment]
@demobox 

> modify the "copy stage" behaviour to (perhaps optionally) not copy any derived/generated properties. This seems to be the root cause of the confusion here - a property is being duplicated that the user isn't aware of just looking at the UI, and so is probably not expecting to be copied

this is my take on it and just 2 minutes ago I had a member, from another team, tell me he was experiencing the same artifact id's for all of his `Find Artifact from Execution` stages. Turns out he'd gone down the `Copy Stage` path to save some time, i.e he ran into the exact same thing as I did 😄 

I'd love to see `Find Artifact from Execution` to be able to define multiple artifacts, that would be 😍

**Comment by maggieneterval** [comment]
Hey @thecodejunkie, just merged Deck and Orca PRs to support selecting multiple artifacts -- would love your feedback! This will be released with 1.15 (but is of course available now if you're willing to build off the master-unvalidated BOM) :smile:
